File: 329-longest-increasing-path-in-a-matrix/longest-increasing-path-in-a-matrix.go
func inBounds(i, j, numRows, numCols int) bool {
    return i >= 0 && i < numRows && j >= 0 && j < numCols
}

func longestIncreasingPath(matrix [][]int) int {
    n, m := len(matrix), len(matrix[0])
    memo := make([][]int, n)
    for i := range memo{
        memo[i] = make([]int, m)
        for j := range memo[i]{
            memo[i][j] = -1
        }
    }

    var dfs func(i, j int) int
    dfs = func(i, j int) int {
        if memo[i][j] != -1 {
            return memo[i][j]
        }
        longestLen := 1

        for _, offset := range [][2]int{{0, -1}, {0, 1}, {-1, 0}, {1, 0}}{
            new_i := i + offset[0]
            new_j := j + offset[1]
            if inBounds(new_i, new_j, n, m) && matrix[new_i][new_j] > matrix[i][j] {
                longestLen = max(longestLen, dfs(new_i, new_j) + 1)
            }
        }
        memo[i][j] = longestLen
        return longestLen
    }

    maxLen := 1
    for i := range n {
        for j := range m {
            maxLen = max(maxLen, dfs(i, j))
        }
    }

    return maxLen
}

File: 329-longest-increasing-path-in-a-matrix/longest-increasing-path-in-a-matrix.py
class Solution:
    def longestIncreasingPath(self, grid: List[List[int]]) -> int:
        """
        Approach:
        - maxdepth = 1, memo = {}: stores (i, j) -> max length
        - for each node:
            do DFS
        dfs(node):
          if in memo: return value in memo
          
          maxdepth = 0
          for nbr:
            maxdepth = max(maxdepth, dfs(nbr))
          return 1 + maxdepth
        """
        n, m = len(grid), len(grid[0])
        memo = {} # store (row, col) -> depth

        def dfs(row: int, col: int) -> None:
            if (row, col) in memo:
                return memo[(row, col)]
            
            max_depth = 0
            for offset in [[-1, 0], [1, 0], [0, -1], [0, 1]]:
                new_row = row + offset[0]
                new_col = col + offset[1]

                if 0 <= new_row < n and 0 <= new_col < m and grid[new_row][new_col] > grid[row][col]:
                    max_depth = max(max_depth, dfs(new_row, new_col))
            memo[(row, col)] = 1 + max_depth
            return memo[(row, col)]
        
        longest_path = 1
        for i in range(n):
            for j in range(m):
                longest_path = max(longest_path, dfs(i, j))
        
        return longest_path